stdlib/asm/collections/mmr.masm

```diff
@@ -0,0 +1,81 @@
+#! Computes the `ilog2(number)` and `2^(ilog2(number))`.
+#!
+#! number must be non-zero, otherwise this will error
+#!
+#! Stack transition:
+#! [number, ...] -> [ilog2, power_of_two, ...]
+export.ilog2_checked
+  push.2147483648 # power_of_two from high to low bit
+  push.0          # bit_pos from the most-signficant, `31-bit_pos` equals to ilog2
+  # stack: [bit_pos, power_of_two, number, ...]
+
+  # find the first most-significant true bit
+  dup.1 dup.3 u32checked_and eq.0 while.true
+    add.1 swap div.2 swap
+    dup.1 dup.3 u32checked_and eq.0
+  end
+
+  # compute ilog2
+  push.31 swap sub
+  # stack: [ilog2, power_of_two, number, ...]
+
+  # drop number
+  movup.2 drop
+  # stack: [ilog2, power_of_two, ...]
+end
+
+#! Loads the leaf at the absolute `pos` in the MMR.
+#!
+#! Stack transition:
+#! [pos, mmr_ptr, ...] -> [n3, n2, n1, n0, ...]
+export.get
+  # load the num_leaves of the MMR
+  dup.1 mem_load
+  # stack: [num_leaves, pos, mmr_ptr, ...]
+
+  # compute `num_leaves & pos`, this contains all peaks before `pos` (and maybe some after the owning peak)
+  dup.1 dup.1 u32checked_and
+  # stack: [before_candidates, num_leaves, pos, mmr_ptr, ...]
+
+  # compute `num_leaves - before_candidates`, this removes every peak before the owner (result may include peaks after owner)
+  dup.1 swap sub
+  # stack: [owner_candidates, num_leaves, pos, mmr_ptr, ...]
+
+  # compute `ilog2(owner_candidates)` and `2**depth`, it corresponds to the owner peak and its depth
+  exec.ilog2_checked swap
+  # stack: [owner_peak, depth, num_leaves, pos, mmr_ptr, ...]
+
+  # compute `owner_peak - 1`, this mask corresponds to every peak after the owner
+  dup.0 sub.1
+  # stack: [after_mask, owner_peak, depth, num_leaves, pos, mmr_ptr, ...]
+
+  # compute `num_leaves & after_mask`, uses the mask to compute the actual after peaks
+  dup.3 u32checked_and
+  # stack: [after_peaks, owner_peak, depth, num_leaves, pos, mmr_ptr, ...]
+
+  # compute `num_leaves - (after_peaks + owner_peak)`, this computes the before peaks
+  add movup.2 swap sub
+  # stack: [peaks_before, depth, pos, mmr_ptr, ...]
+
+  # compute `pos - peaks_before`, this computes the relative_pos of the leaf w.r.t. the owner peak.
+  movup.2 dup.1 sub
+  # stack: [relative_pos, peaks_before, depth, mmr_ptr, ...]
+
+  # compute `popcount(peaks_before)`, the count peaks before the target to be skipped when loading from mem
+  swap u32checked_popcnt
+  # stack: [peak_count, relative_pos, depth, mmr_ptr, ...]
+
+  # compute `mmr_ptr + peak_count + 1` the target tree index
+  movup.3 add add.1
+  # stack: [mmr_ptr, relative_pos, depth, ...]
+
+  # load the target peak
+  padw movup.4 mem_loadw
+  # stack: [P, relative_pos, depth, ...]
+
+  # find the tree depth
+  movup.4 movup.5
+  # stack: [depth, relative_pos, P, ...]
+
+  mtree_get
```

Comment:
How would this work if the `owner_peak` is the last node and this node is not a tree (but just a single leave). We won't be able to use `mtree_get` in this situation, right?

Reply:
ah, good call, the element would not be in the merkle store given the current implementation. I will fix it by returning `P` if depth is `0`.